android/PhysicalWeb/app/src/main/java/org/physical_web/physicalweb/WifiUrlDeviceDiscoverer.java

```diff
@@ -0,0 +1,122 @@
+ */
+class WifiUrlDeviceDiscoverer extends UrlDeviceDiscoverer {
+  private final String TAG = "WifiUrlDeviceDiscoverer";
+  private final Context mContext;
+  private final WifiP2pManager mManager;
+  private final Channel mChannel;
+  private WiFiDirectBroadcastReceiver mReceiver;
+  private final IntentFilter mIntentFilter = new IntentFilter();
+  private final PeerListListener myPeerListListener = new PeerListListener() {
+    @Override
+    public void onPeersAvailable(WifiP2pDeviceList list) {
+      Log.d(TAG, list.toString());
+      for (WifiP2pDevice device : list.getDeviceList()) {
+        Utils.WifiDirectInfo info = Utils.parseWifiDirectName(device.deviceName);
+        if (info != null) {
+          String name = info.title;
+          int port = info.port;
+          reportUrlDevice(createUrlDeviceBuilder("WifiDirect" + name,
+              device.deviceAddress + ":" + port)
+            .setWifiAddress(device.deviceAddress)
+            .setWifiPort(port)
+            .setTitle(name)
+            .setDescription("")
+            .build());
+        }
+      }
+    }
+  };
+
+  public WifiUrlDeviceDiscoverer(Context context) {
+    mManager = (WifiP2pManager) context.getSystemService(Context.WIFI_P2P_SERVICE);
+    mChannel = mManager.initialize(context, context.getMainLooper(), null);
+    mContext = context;
+    mReceiver = new WiFiDirectBroadcastReceiver(mManager, mChannel);
+  }
+
+  @Override
+  public synchronized void startScanImpl() {
+    IntentFilter intentFilter = new IntentFilter();
+    intentFilter.addAction(WifiP2pManager.WIFI_P2P_PEERS_CHANGED_ACTION);
+    mContext.registerReceiver(mReceiver, intentFilter);
+    mManager.discoverPeers(mChannel, new WifiP2pManager.ActionListener() {
+      @Override
+      public void onSuccess() {
+        Log.d(TAG, "discovering");
+      }
+
+      @Override
+      public void onFailure(int reasonCode) {
+        Log.d(TAG, "discovery failed " + reasonCode);
+      }
+    });
+  }
+
+  @Override
+  public synchronized void stopScanImpl() {
+    mManager.stopPeerDiscovery(mChannel, new WifiP2pManager.ActionListener() {
+      @Override
+      public void onSuccess() {
+        Log.d(TAG, "stop discovering");
+      }
+
+      @Override
+      public void onFailure(int reasonCode) {
+        Log.d(TAG, "stop discovery failed " + reasonCode);
+      }
+    });
+  }
+
+  public class WiFiDirectBroadcastReceiver extends BroadcastReceiver {
+
+    private WifiP2pManager manager;
+    private Channel channel;
+
+    public WiFiDirectBroadcastReceiver(WifiP2pManager manager, Channel channel) {
+      super();
+      this.manager = manager;
+      this.channel = channel;
+    }
+
+    @Override
+    public void onReceive(Context context, Intent intent) {
+      String action = intent.getAction();
+      if (WifiP2pManager.WIFI_P2P_PEERS_CHANGED_ACTION.equals(action)
+          && mManager != null) {
+        mManager.requestPeers(mChannel, myPeerListListener);
```

Comment:
The constructor assigns this.manager and this.channel, but here we're using the parent class's mManager and mChannel.  It's fine either way, but please remove the unused members.  I think this receiver would also work well as an anonymous class.